src/org/rascalmpl/util/maven/Project.java

```diff
@@ -0,0 +1,274 @@
+            }
+            var message = problem.getMessage();
+            switch (problem.getSeverity()) {
+                case ERROR: // fall through
+                case FATAL: messages.append(Messages.error(message, pos)); break;
+                case WARNING: messages.append(Messages.warning(message, pos)); break;
+                default: throw new UnsupportedOperationException("Missing case: " + problem.getSeverity());
+            }
+        }
+        return messages;
+    }
+
+    private static Project buildError(ModelBuildingException be, CurrentResolution context,  boolean isRoot) {
+        var messages = translateProblems(be.getProblems(), context.pom);
+        var incompleteModel = be.getModel();
+        if (incompleteModel == null) {
+            messages.append(Messages.error("Could not build an intermediate model", context.pom));
+            return new Project(ArtifactCoordinate.UNKNOWN, null, Collections.emptyList(), messages.done(), true);
+        }
+        return translate(incompleteModel, messages, true, context, isRoot);
+    }
+
+    private static ArtifactCoordinate translateCoordinate(Model model) {
+        return new ArtifactCoordinate(model.getGroupId(), model.getArtifactId(), model.getVersion());
+    }
+
+    private static ArtifactCoordinate translateCoordinate(Parent model) {
+        if (model == null || model.getArtifactId() == null || model.getArtifactId().isEmpty()) {
+            return null;
+        }
+        return new ArtifactCoordinate(model.getGroupId(), model.getArtifactId(), model.getVersion());
+    }
+
+    private static List<Dependency> translateDependencies(Model model, IListWriter messages, CurrentResolution context, boolean isRoot) {
+        return model.getDependencies()
+            .stream()
+            .filter(d -> !"system".equals(d.getScope())) // we don't care about system deps
+            .filter(d -> isRoot || !"test".equals(d.getScope())) // unless we're the root, we don't care about downstream test dependencies
+            .map(d -> Dependency.build(d, messages, context))
+            .collect(Collectors.toUnmodifiableList())
+            ;
+    }
+
+    @Override
+    public String toString() {
+        var result = new StringWriter();
+        try (var printer = new PrintWriter(result)) {
+            write(printer);
+        }
+        return result.toString();
+    }
+
+    public void write(PrintWriter target) {
+        target.print("Maven project: ");
+        target.println(coordinate);
+        target.print("Parent: ");
+        target.println(parentCoordinate);
+        target.println("Dependencies:");
+        for (var d : dependencies) {
+            writeDependencies(target, "", d);
+        }
+        target.println("Messages:");
+        Messages.write(messages, target);
+    }
+
+
+    private void writeDependencies(PrintWriter target, String prefix, Dependency d) {
+        target.print(prefix);
+        target.print("- ");
```

Comment:
Better use -- since - is used a lot inside the prefix and inside an artifact Id

Reply:
This is just for the indented print of a tree of dependencies. Nothing fancy needed.